### PR TITLE
Api week1 smith

### PR DESCRIPTION
**Pull request by asksmith**

### CODE REVIEW

**Review by remarcmij — CHANGES_REQUESTED** [review]
Hi @asksmith, good work (but please note my message in the class39 channel about the old homework).

Please fix the omission in ex1. Other than that just some suggestions. Also, check your Prettier setup.

File: 3-UsingAPIs/Week1/homework/ex4-pokerDiceAll.js [hunk 1]
```diff
@@ -27,9 +27,12 @@ exercise file.
...
 
 function rollDice() {
-  // TODO Refactor this function
   const dice = [1, 2, 3, 4, 5];
-  return rollDie(1);
+  return Promise.all(
+    dice.map((item) => {
+      return rollDie(item);
+    })
+  );
```
[the remark this answers is not in the document]
Comment (remarcmij) [reply 3-UsingAPIs/Week1/homework/ex4-pokerDiceAll.js:35]: I'm missing the requested explanation:

- You may also notice that, in the case of a rejected promise, dice that have not yet finished their roll continue to do so. Can you explain why? Please add your answer as a comment to the end of the exercise file.

**Comment by asksmith** [comment]
I've fix all up

File: 3-UsingAPIs/Week1/homework/ex3-rollDie.js [hunk 2]
```diff
@@ -36,21 +36,29 @@ function rollDie() {
...
 }
 
 // ! Do not change or remove the code below
 if (process.env.NODE_ENV !== 'test') {
   main();
 }
 module.exports = rollDie;
+/*
+ **  With the callback version if any error occurs it will continue but with the Promise it will stop when first error occurs!
+ **
```
Comment (remarcmij) [inline 3-UsingAPIs/Week1/homework/ex3-rollDie.js:63]: That's what you observe, but not an explanation why this happens. The key point here is that the first call to either `resolve()` or `reject()` settles the promise to its final value. A subsequent call to either `resolve()` or `reject()` will not change the settled outcome.

File: 3-UsingAPIs/Week1/homework/ex5-pokerDiceChain.js [hunk 3]
```diff
@@ -14,35 +14,15 @@ to expand the given promise chain to include five dice.
...
-      return results;
+      return rollDice(index + 1, results);
     });
-}
-
-function main() {
-  rollDice()
-    .then((results) => console.log('Resolved!', results))
-    .catch((error) => console.log('Rejected!', error.message));
+  }
```
Comment (remarcmij) [inline 3-UsingAPIs/Week1/homework/ex5-pokerDiceChain.js:25]: While the test succeeds your code does not run when used by itself. In that case your `rollDice()` function is never called (you removed the `main()` function).

Your solution should work with the original `main()` function when executed directly (not as a test).

File: 3-UsingAPIs/Week1/homework/ex5-pokerDiceChain.js [hunk 4]
```diff
@@ -14,35 +14,15 @@ to expand the given promise chain to include five dice.
...
 const rollDie = require('../../helpers/pokerDiceRoller');
 
-function rollDice() {
-  const results = [];
-  return rollDie(1)
-    .then((value) => {
+function rollDice(index = 0, results = []) {
+  if (index > 4) {
+    return results;
+  } else {
```
Comment (remarcmij) [inline 3-UsingAPIs/Week1/homework/ex5-pokerDiceChain.js:20]: While I commend you for your clever recursive solution it at the same time requires me to do a "double (or) triple take": I need to go over your code several times to understand how it works. Luckily `async/await` can eliminate the need for recursion like this (week 2 stuff).

Note that you don't need an `else` after a return.

**Review by remarcmij — APPROVED** [review]
Hi @asksmith, well done, I can approve your PR now. I still have some comments but no need for a new review round.

Good luck with the project!

**Pull request closed without merging**